Codes/Clean Data/getRealizedVolatility.R

```diff
@@ -0,0 +1,84 @@
+function(n,
+         delta,
+         K,
+         ss,
+         intervals,
+         q,
+         time,
+         mid,
+         string1) {
+  
+  
+  mid <- log(mid) #take log midquotes
+  
+  if (!string1 %in% c("PRE", "POST")) {
+    stop("String must be either PRE or POST")
+  }
+  
+  if (string1 == "PRE") {
+    RV_full <- matrix(NA, 1, K)
+    
+    for (xx in 1:K) {
+      intervaltemp <-
+        seq(intervals[q - n] - (xx - 1) * ss, intervals[q + 1] - (xx - 1) * ss, delta) #split this interval into n intervals
+      
+      RV <- matrix(NA, 1, n)
+      
+      for (ii in 1:(length(intervaltemp) - 1)) {
+        iia <- min(which(time > intervaltemp[ii]))

```

Comment:
Every time you use which, R has to scan all time data again. You want to get is the previous midquote at each interval. You can indeed create a data.table with regular time interval and do as-of join on quotes data.table.